Automattic/wp-calypso
# Design picker: Include Premium Themes in Personal & Free Plans with checkout flow

#### Changes proposed in this Pull Request

* Make an `<AsyncCheckoutModal />` so that the user can complete the checkout on the current page and won't be redirected to the checkout page and load a bunch of files.
* Use `signup/design-picker-premium-themes-checkout` to control the user with free or personal plan can see premium themes in Design Picker or not
* Check the premium themes are available to the user. If not, it would show the “Upgrade Plan” when the user hovers on the theme or preview the theme.
* Persist exclude steps for the flow with branching steps. As the Design Picker has to checkout for premium theme and the user might be redirected to 3rd-party payment such as PayPal, the branching steps won't work after they come back from 3rd-party payment.

Note that if you test with the local calypso, it might not work after you come back from 3rd-party payment as the signup state might be clear on the development environment. See https://github.com/Automattic/wp-calypso/blob/f298c04c540123c453a7b951550fe7321f394734/client/state/initial-state.js#L60

https://user-images.githubusercontent.com/13596067/150091494-4da54d3a-05e8-476e-89fc-b7414dd154ca.mov

#### Testing instructions



* Go to `/start/setup-site?siteSlug=<your_site>` with free plan or personal plan
* When you land on the Design Picker, you should see the premium themes with “Upgrade Plan”
* Click on the “Upgrade Plan” button, the checkout modal will open, and the annual premium plan will be already there.
* After you complete the checkout, the modal would be closed and you can see “Start with <name of premium theme>“ now!
* If you close the checkout modal or remove the plan, the modal will be closed but the text of the button will still be “Upgrade Plan”



Related to https://github.com/Automattic/wp-calypso/issues/59229
